src/ModelContextProtocol/Protocol/Transport/StreamableHttpClientSessionTransport.cs

```diff
@@ -0,0 +1,267 @@
+
+/// <summary>
+/// The Streamable HTTP client transport implementation
+/// </summary>
+internal sealed partial class StreamableHttpClientSessionTransport : TransportBase
+{
+    private static MediaTypeWithQualityHeaderValue ApplicationJsonMediaType = new("application/json");
+    private static MediaTypeWithQualityHeaderValue TextEventStreamMediaType = new("text/event-stream");
+
+    private readonly HttpClient _httpClient;
+    private readonly SseClientTransportOptions _options;
+    private readonly CancellationTokenSource _connectionCts;
+    private readonly ILogger _logger;
+
+    private string? _mcpSessionId;
+    private Task? _getReceiveTask;
+
+    public StreamableHttpClientSessionTransport(SseClientTransportOptions transportOptions, HttpClient httpClient, ILoggerFactory? loggerFactory, string endpointName)
+        : base(endpointName, loggerFactory)
+    {
+        Throw.IfNull(transportOptions);
+        Throw.IfNull(httpClient);
+
+        _options = transportOptions;
+        _httpClient = httpClient;
+        _connectionCts = new CancellationTokenSource();
+        _logger = (ILogger?)loggerFactory?.CreateLogger<SseClientTransport>() ?? NullLogger.Instance;
+
+        // We connect with the initialization request with the MCP transport. This means that any errors won't be observed
+        // until the first call to SendMessageAsync. Fortunately, that happens internally in McpClientFactory.ConnectAsync
+        // so we still throw any connection-related Exceptions from there and never expose a pre-connected client to the user.
+        SetConnected(true);
+    }
+
+    /// <inheritdoc/>
+    public override async Task SendMessageAsync(
+        JsonRpcMessage message,
+        CancellationToken cancellationToken = default)
+    {
+        using var sendCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _connectionCts.Token);
+        cancellationToken = sendCts.Token;
+
+        using var content = new StringContent(
+            JsonSerializer.Serialize(message, McpJsonUtilities.JsonContext.Default.JsonRpcMessage),
+            Encoding.UTF8,
+            "application/json"
+        );
+
+        using var httpRequestMessage = new HttpRequestMessage(HttpMethod.Post, _options.Endpoint)
+        {
+            Content = content,
+            Headers =
+            {
+                Accept = { ApplicationJsonMediaType, TextEventStreamMediaType },
+            },
+        };
+
+        CopyAdditionalHeaders(httpRequestMessage.Headers, _options.AdditionalHeaders, _mcpSessionId);
+        var response = await _httpClient.SendAsync(httpRequestMessage, HttpCompletionOption.ResponseHeadersRead, cancellationToken).ConfigureAwait(false);
+
+        response.EnsureSuccessStatusCode();
+
+        var rpcRequest = message as JsonRpcRequest;
+        JsonRpcMessage? rpcResponseCandidate = null;
+
+        if (response.Content.Headers.ContentType?.MediaType == "application/json")
```

Comment:
Do we need to care about situations where the media type contains char sets or other optional data that would make this not compare equally even when json?

Reply:
This is reading from the pre-parsed MediaTypeHeaderValue which already spits out the encoding and strips away any extra whitespace or other parameters like the quality factor.